Cleanup RunningAverage implementations

bertm opened the pull request:
the RunningAverage implementations have various issues:
* Too many responsibilities (like TimeDecayingRunningAverage, that tries to keep track of clock skew in the middle of its calculations)
* Too much mutable state (like SimpleRunningAverage, that keeps track of "how many items do I have" across 3 different fields)
* A generous amount of synchronization, but not everywhere
* Leftover methods from earlier times (custom serialization formats and unused code that does who-knows-what)
* A large amount of empty Javadoc comment
* A general lack of consistency in code formatting
* Zero tests

This PR attempts to improve the situation by:
* Separating out the clock skew tracking & reporting out of the main logic of TimeDecayingRunningAverage (and using a monotonic clock for the actual computations)
* Keeping the state in a single immutable object accessed through an AtomicReference where possible, so state is in one place and can easily be reasoned about
* Replacing synchronization with compare-and-set operations on the reference
* Removing the legacy format read and write method, and removing Serializable from the class hierarchy (this is technically a breaking, but in my view it is highly unlikely that anybody would be using these methods)
* Removing the empty Javadoc. Most of the operations are explained on the RunningAverage interface; others are sufficiently clear.
* Reformat everything (consistent use of tabs vs spaces, yay!), move constructors to the top, etc.
* Add some tests to cover most of the functionality

Also, deprecate MedianMeanRunningAverage as it's probably never a good idea to use it.

ArneBab commented on test/freenet/support/math/BootstrappingDecayingRunningAverageTest.java:
```diff
@@ -0,0 +1,93 @@
+        average.report(Double.NEGATIVE_INFINITY);
+        average.report(Double.POSITIVE_INFINITY);
+        average.report(Double.NaN);
+        assertThat(average.countReports(), equalTo(1L));
```
also check that the value is 0?

bertm replied:
Fair enough, added an assertion for that!

ArneBab commented on src/freenet/support/math/TrivialRunningAverage.java:
```diff
@@ -1,77 +1,69 @@
+        }
+
+        Data updated(double d) {
+            return new Data(reports + 1, d + total);
```
This seems very expensive. Do we really need it?

bertm replied:
This adds two numbers and constructs a tiny object. What exactly seems very expensive here? :)

> Do we really need the trivial running average?

It is used all over the place (46 locations) so we cannot just remove it.

ArneBab commented on src/freenet/support/math/MedianMeanRunningAverage.java:
```diff
@@ -1,98 +1,80 @@
+        synchronized (reports) {
+            int size = reports.size();
+            int middle = size / 2;
+            Collections.sort(reports);
```
Is the default sort optimal for already mostly sorted data?

Yes, it is:

> If the input array is nearly sorted, the implementation requires approximately n comparisons https://docs.oracle.com/javase/8/docs/api/java/util/List.html#sort-java.util.Comparator-

ArneBab commented on src/freenet/node/RequestSender.java:
```diff
@@ -1542,9 +1543,9 @@ public synchronized short waitUntilStatusChange(short mask) {
+	private static final RunningAverage avgTimeTaken = new TrivialRunningAverage();
 	
-	private static MedianMeanRunningAverage avgTimeTakenTransfer = new MedianMeanRunningAverage();
+	private static final RunningAverage avgTimeTakenTransfer = new TrivialRunningAverage();
```
Why does this use the TrivialRunningAverage, when there are much more efficient algorithms? This looks like a good case for the now more efficient `BootstrappingDecayingRunningAverage`.

bertm replied:
MedianMeanRunningAverage is expensive (and is therefore deprecated) as it has O(1) updates but O(n log n) queries and O(n) memory usage.

TrivialRunningAverage is cheap with O(1) updates, O(1) queries and O(1) memory usage.

BootstrappingDecayingRunningAverage might give slightly more useful metrics, but note that `avgTimeTakenTransfer` is only used when `logMINOR` is enabled for this class - this is mostly a debugging tool that probably nobody actually uses.

bertm commented:
@ArneBab thank you for your review!

Addressed your comment regarding the additional assertion and rebased this branch on the latest `next`.
Hope I cleared up the `TrivialRunningAverage` discussions too - just let me know if we're on the same page now :smile:

ArneBab commented:
merged -- thank you!
